plat/kvm/arm/smp.c

```diff
@@ -0,0 +1,402 @@
+
+			barrier();
+
+			err = cpu_on(&(cpus[cpu_idx]), lcpu_start_addr);
+			if (err != PSCI_RET_SUCCESS) {
+				uk_pr_err("Failed to start CPU %ld\n", cpus[cpu_idx].id, err);
+				cpus[cpu_idx].state = LCPU_STATE_OFFLINE;
+				ukarch_spin_unlock(&(cpus[cpu_idx].lock));
+				continue;
+			}
+
+			uk_pr_info("Started core (%ld) successfully\n",
+				cpus[cpu_idx].id);
+			ukarch_spin_unlock(&cpus[cpu_idx].lock);
+			n++;
+		}
+	}
+
+	return n;
+}
+
+void __noreturn lcpu_entry_default(void)
+{
+	intctrl_init();
+	ukplat_lcpu_enable_irq();
+
+	/* Enter halt loop. Besides interrupts in general, the halt can be
+	 * interrupted by calls to ukplat_lcpu_run().
+	 */
+	while (1) {
+		cpus[ukplat_lcpu_idx()].state = LCPU_STATE_IDLE;
+		halt();
+	}
+}
+
+int ukplat_lcpu_run(__lcpuid lcpuid[], unsigned int num,
+		struct ukplat_lcpu_func *fn, int flags)
+{
+	__u32 i;
+	__u64 this_cpu_id = ukplat_lcpu_id();
+	__lcpuid cpu_idx;
+
+	UK_ASSERT(fn);
+
+	if (num == 0) {
+		UK_ASSERT(!lcpuid);
+		num = smp_numcores;
+	}
+
+	UK_ASSERT(num <= smp_numcores);
+
+	for (i = 0; i < num; i++) {
+		cpu_idx = (lcpuid) ? lcpuid[i] : (__lcpuid) i;
+
+		if (cpus[cpu_idx].id == this_cpu_id)
+			continue;
+
+		ukarch_spin_lock(&cpus[cpu_idx].lock);
+		uk_list_add(&fn->lentry, &cpus[cpu_idx].fnlist);
+		intctrl_send_ipi(UKPLAT_ARM64_IPI_RUN_IRQ, cpus[cpu_idx].id);
+	}
+
+	return 0;
+}
+
+int ukplat_lcpu_wait(__lcpuid lcpuid[], unsigned int num, __nsec timeout)
+{
+	__u32 i;
+	__u64 this_cpu_id = ukplat_lcpu_id();
+	__lcpuid cpu_idx;
+	__nsec end;
+
+	if (num == 0) {
+		UK_ASSERT(!lcpuid);
+		num = smp_numcores;
+	}
+	if (num > smp_numcores) {
+		uk_pr_warn("More cores are selected, but not available\n");
+		num = smp_numcores;
+	}
+
+	if (timeout > 0)
+		end = ukplat_monotonic_clock() + timeout;
+
+	for (i = 0; i < num; i++) {
+		cpu_idx = (lcpuid) ? lcpuid[i] : (__lcpuid) i;
+
+		if (cpus[cpu_idx].id == this_cpu_id)
+			continue;
+
+		ukarch_spin_lock(&cpus[cpu_idx].lock);
+		while (cpus[cpu_idx].state != LCPU_STATE_IDLE)
+			if (timeout && (ukplat_monotonic_clock() > end)) {
+				ukarch_spin_unlock(&cpus[cpu_idx].lock);
+				return -ECANCELED;
+			}
+		ukarch_spin_unlock(&cpus[cpu_idx].lock);
+	}
+
+	return 0;
+}
+
+int ukplat_lcpu_wakeup(__lcpuid	lcpuid[], unsigned int num)
+{
+	__u32 i;
+	__u64 this_cpu_id = ukplat_lcpu_id();
+	__lcpuid cpu_idx;
+
+	if (num == 0) {
+		UK_ASSERT(!lcpuid);
+		num = smp_numcores;
+	}
+	if (num > smp_numcores) {
+		uk_pr_warn("More cores are selected, but not available\n");
+		num = smp_numcores;
+	}
+
+	for (i = 0; i < num; i++) {
+		cpu_idx = (lcpuid) ? lcpuid[i] : (__lcpuid) i;
+
+		if (cpus[cpu_idx].id == this_cpu_id)
+			continue;
+
+		intctrl_send_ipi(UKPLAT_ARM64_IPI_WAKEUP_IRQ, cpus[cpu_idx].id);
+	}
+
+	return 0;
+}
+
+__u64 ukplat_lcpu_id(void)
+{
+	__u64 mpidr_reg;
+
+	mpidr_reg = SYSREG_READ64(mpidr_el1);
+
+	/* return the affinity bits for the current core */
+	return mpidr_reg & CPU_ID_MASK;
```

Comment:
From what I knew, the id of the core is composed by the affinity bits in MPIDR_EL1. The other bits are giving information about the system (e.g. single or multi-core) or are simply reserved: https://developer.arm.com/documentation/ddi0601/2021-12/AArch64-Registers/MPIDR-EL1--Multiprocessor-Affinity-Register

Reply:
There is more information there, that tells how to interpret the affinity bits (ie the MT bit). Moreover, I don't see this any information that states that the core ID consists of specific parts of `MPIDR_EL1`. In fact, the values of the affinity bits are are IMPLEMENTATION DEFINED. The exact details are described in each core's TRM (Technical Reference Manual).

My point here is, why try to guess the caller's intentions? By stripping the semantical information about affinity, you now return an opaque value. Sure, it will serve as a unique identifier, but for any other purpose where the caller needs more information, they will have no choice but issuing an `MRS` to get the real value of `MPIDR`.